### ISSUE TITLE
Formatting not working with neovim when rubocop is a transient dependency

### ISSUE THREAD

**Issue by dgmora** [Shopify/ruby-lsp#1198]
### Ruby version

3.2.2

### Description

I have followed the intructions for setting ruby-lsp with neovim, and it worked correctly it has stopped working. I've tried going back to a previous version but no luck. Auto-formatting works for me with other lsps, so that's why I'm looking into ruby-lsp. Diagnostics work though

Here's my example class:

```ruby
1  | class User < ApplicationRecord
      . . .
24 |   def foo
25 |      puts "foo"
26 |    end
27 |    
28 |    private # This line has been indented incorrectly. There is a diagnostic
29 |    
30 |   def bar
31 |      puts "bar"
32 |   end
      . . .
```

<details><summary>I have checked the logs, and this is what I'm getting after saving the change to line 28:</summary>

```
[DEBUG][2023-11-22 10:14:03] .../lua/vim/lsp.lua:1487 "LSP[ruby_ls]"  "client.request"  1 "textDocument/documentHighlight"  {
  position = {
    character = 4,
    line = 27
  },
  textDocument = {
    uri = "file:///Users/david/src/github.com/dgmora/my-project/app/models/user.rb"
  }
} <function 1>  1
[DEBUG][2023-11-22 10:14:03] .../vim/lsp/rpc.lua:261  "rpc.send"  {
  id = 231,
  jsonrpc = "2.0",
  method = "textDocument/documentHighlight",
  params = {
    position = {
      character = 4,
      line = 27
    },
    textDocument = {
      uri = "file:///Users/david/src/github.com/dgmora/my-project/app/models/user.rb"
    }
  }
}
[DEBUG][2023-11-22 10:14:03] .../lua/vim/lsp.lua:1487 "LSP[ruby_ls]"  "client.request"  1 "textDocument/formatting" {
  options = {
    insertSpaces = true,
    tabSize = 2
  },
  textDocument = {
    uri = "file:///Users/david/src/github.com/dgmora/my-project/app/models/user.rb"
  }
} <function 1>  1
[DEBUG][2023-11-22 10:14:03] .../vim/lsp/rpc.lua:261  "rpc.send"  {
  id = 232,
  jsonrpc = "2.0",
  method = "textDocument/formatting",
  params = {
    options = {
      insertSpaces = true,
      tabSize = 2
    },
    textDocument = {
      uri = "file:///Users/david/src/github.com/dgmora/my-project/app/models/user.rb"
    }
  }
}
[DEBUG][2023-11-22 10:14:03] .../vim/lsp/rpc.lua:364  "rpc.receive" {
  id = 231,
  jsonrpc = "2.0",
  result = { {
      kind = 2,
      range = {
        ["end"] = {
          character = 11,
          line = 27
        },
        start = {
          character = 4,
          line = 27
        }
      }
    } }
}
[DEBUG][2023-11-22 10:14:03] .../vim/lsp/rpc.lua:364  "rpc.receive" {
  id = 232,
  jsonrpc = "2.0"
}
[DEBUG][2023-11-22 10:14:03] .../vim/lsp/rpc.lua:261  "rpc.send"  {
  jsonrpc = "2.0",
  method = "workspace/didChangeWatchedFiles",
  params = {
    changes = { {
        type = 2,
        uri = "file:///Users/david/src/github.com/dgmora/my-project/app/models/user.rb"
      } }
  }
}

```

</details>

I'm not sure if this helps to understand if something doesn't look right.  I see a request for `textDocument/formatting`, not sure if there should be something else.

### Expected output

Formatting should work

**Comment by vinistock** [comment]
Thank you for the bug report! Are you using `rubocop` and is the gem in your Gemfile? Are you passing any initialization options in your LSP config?

Also, are you using the LSP in a symlink folder?

**Comment by waiting-for-dev** [comment]
Not opening a new issue, as I'd need more investigation. In my case, rubocop is working, but it looks like the `.rubocop.yml` file is ignored and it runs through a vanilla rubocop configuration.

**Comment by rafaelrinaldi** [comment]
@dgmora Can you please share your configuration? I cannot get auto formatting to work.

**Comment by vinistock** [comment]
> The problem in this case is that I don't have rubocop as a direct dependency

This is by design. We try to automatically detect which formatter your project is using, but having it listed as a direct dependency in the `Gemfile` is required.

The reason is because checking for transitive dependencies can lead to awkward situations. For example, Standard has a dependency on RuboCop, which makes it more difficult to determine what formatter the project is actually using.

Furthermore, dependabot doesn't open upgrade PRs for transitive dependencies, so having your formatter in the `Gemfile` ensures that you're getting the improvements from newer versions.

TL;DR: this is by design, please add your formatter as a direct dependency.

**Comment by vinistock** [comment]
> Using rubocop as a transient dependency is something that rubocop allows by design, so not supporting that in some way feels awkward.

This isn't RuboCop specific. Any transitive dependency can be used directly by a project. However, using a transitive dependency creates an implicit dependency on it.

In general, I would argue it's not good practice. What if the gem that's bringing the dependency that your app depends on decides to drop it? Suddenly your app could break because you depend on something that's not explicitly listed as a dependency.

> Could you be more specific about those awkward situations? This example for instance shouldn't be difficult to support

Supporting this particular example is easy, but any other gem can depend on a formatter. For example, `syntax_tree` includes a parser and a formatter, which means any gem depending on it would bring another formatter into the transitive dependencies.

Trying to account for all possible cases where a formatter is available through transitive dependencies would not scale well. It would require us to build custom rules depending on other gems.

On the other hand, recommending that your preferred formatter be a part of the Gemfile scales easily. It's very unlikely that an application will have a direct dependency on two or more formatters, so it makes the automatic detection reliable.

I should also mention that there is a workaround. If you truly can't add `rubocop` as a direct dependency, you can pass `formatter: "rubocop"` as an initialization option to the LSP and it will bypass automatic detection.

> I don't think it'll be easy to find out for everyone

This is fair. How about we print a warning if RuboCop is found, but not in the direct dependencies?

**Comment by vinistock** [comment]
@dgmora thank you for the thoughtful discussion. I'll bring this up with the team and think about we could do.

**Comment by dgmora** [comment]
Happy to help, thanks for the great work you do for the community 🙂 

By chance I've seen a recent message in slack that also works around the issue: Passing the `formatter=rubocop` option when starting the lsp. The issue only occurrs with `auto`. In my case this would be `init_options = { formatter = "rubocop" }`.
I think it'd still be nice if it's possible to fix it for `auto`

**Comment by vinistock** [comment]
Okay, we talked among the team and agreed to change the logic to the following:
```
RuboCop is present as direct dependency ||
  (Workspace has .rubocop.yml && Syntax Tree is not a direct dependency)
```

This is necessary because [Syntax Tree supports usage with RuboCop](https://github.com/ruby-syntax-tree/syntax_tree?tab=readme-ov-file#rubocop), which means just checking for `.rubocop.yml` would also cause an incorrect selection for people using RuboCop + Syntax Tree.

That said, this is not a super straight forward change. Currently, our [DependencyDetector](https://github.com/Shopify/ruby-lsp/blob/main/lib/ruby_lsp/requests/support/dependency_detector.rb) logic does not have access to the workspace URI announced by the client, which we need in order to join paths and check
```ruby
File.join(workspace_uri, ".rubocop.yml")
```

We can't simply use `.rubocop.yml` without the workspace URI because that fails for workspaces that are symlinks (we've had bugs with this in the past).

First, we need to refactor to ensure that the dependency detector has the workspace URI. Then we can change the RuboCop check.

**Comment by andreaswachowski** [comment]
Just to add that I am in the same boat and spend a good few hours today to narrow this down. So yes, it would be great if this either worked out of the box, or a warning would allow the user to know what's going on. Thanks anyway for ruby-lsp which is great!

**Comment by Earlopain** [comment]
👋could you check if #2126 works for you all? You should be able to try it out by adding the following to your gemfile: `gem "ruby-lsp", github: "https://github.com/Shopify/ruby-lsp/pull/2126"`. https://github.com/Shopify/ruby-lsp/blob/main/CONTRIBUTING.md#manually-testing-a-change

**Comment by dgmora** [comment]
Seems to be working for me without specifying in the init options `rubocop`, thanks! If there's anything I'll share
